### PR TITLE
Add intuitive way to open additional traces

### LINKED ISSUE #193: Add intuitive way to open additional traces
Currently, there is no intuitive way to open additional traces if a trace is already open. Without this, users could become frustrated early in their user experience or assume that you can only open one trace at a time. 

Initially, this use case was going to be covered by adding an "Open Trace" option to the File menu #159, but the VS Code extensions API does not allow contributions to the main menu: 
* VS Code menu contribution points [documentation](https://code.visualstudio.com/api/references/contribution-points#contributes.menus)
* [Comment](https://github.com/theia-ide/theia-trace-extension/issues/159#issuecomment-733749083) on Add “Open Trace” option to File menu #159 

The other currently planned options for opening a trace do not cover this use case:
* The "Open Trace" button #160 just covers the case where no traces are open.
* The current alternatives (Right-click in the file explorer, CTRL+SHIFT+P + "Open Trace") would not be the user's first reflex.

## Proposed solution
Keep the "Open Trace" button (from PR #186) in the side panel menu even after opening the first trace (ex. under the opened traces)

image

### PR DESCRIPTION
**Pull request by hriday-panchasara**
fixes #193 

- Button added to the 'Opened Traces' widget header to allow opening additional traces. Button appears when mouse is hovered over widget
- In my approach I'm emitting a new signal when button is clicked, but I can reuse one of the existing signals instead

Named button tooltip "Open New Trace" because I think it differentiates from the context menu "Open Trace" command:
Screen Shot 2021-12-03 at 2 46 48 PM

Screen Shot 2021-12-03 at 2 49 16 PM

Unable to add '+' to the icon because theia commands accept a single iconClass param and I couldn't find a way to stack multiple icons in one element. We can upgrade font-awesome to v5 to get access to a folder-plus icon

### CODE REVIEW

**Review by ebugden — CHANGES_REQUESTED** [review]
Requested changes:

- [x] Change new button tooltip to "Open Trace"
- [ ] Suggestion: Put code related to trace opening in a different file? Rather than having it in the placeholder widget

----

> Button added to the 'Opened Traces' widget header to allow opening additional traces. Button appears when mouse is hovered over widget

* The button interactions match the File Explorer icons perfectly! Good consistency.
* I like this backup icon choice. If we can't have a file with a plus, I think having an open file will make it more clear that this button is for opening traces.

Thank you for these changes @hriday-panchasara! This is one of the most noticeable UX improvements this year! ☀️

**Review by mirsky-work — APPROVED** [review]
Functionality works well! Thanks for this useful contribution!

File: theia-extensions/viewer-prototype/src/browser/trace-explorer/trace-explorer-sub-widgets/trace-explorer-placeholder-widget.tsx
Comment (hriday-panchasara) [reply theia-extensions/viewer-prototype/src/browser/trace-explorer/trace-explorer-sub-widgets/trace-explorer-placeholder-widget.tsx:58]: @PatrickTasse @bhufmann any suggestions? I catch the signal in the placeholder widget because the button would be in that component and the widget already has a method to handle opening of traces. An alternative could be to manage the signal in a parent component like trace-viewer.tsx or trace-explorer-widget.tsx and move handleOpenTrace functionality to a utils file or service

File: theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx
Comment (bhufmann) [inline theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx:95]: you can just call the other open command: registry.executeCommand(OpenTraceCommand.id); No need for the signal.

File: theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-commands.ts
Comment (bhufmann) [inline theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-commands.ts:35]: OPEN_TRACE

File: theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-commands.ts
Comment (bhufmann) [reply theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-commands.ts:37]: I agree, it should be just `Open Trace` since it's the same command than the other one

**Review by bhufmann — CHANGES_REQUESTED** [review]
It can be simplified, no need for all the signal related code.

File: theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx
Comment (bhufmann) [reply theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx:95]: I understand now why the it was moved to the place holder, i.e. to execute the health check as well. So, the command handler of the open command should to the heath check as part of the command execution. That means the code from the placeholder widget has to move into the command-handler. Then the command can be executed from multiple places and the behavior is the same.

Comment (hriday-panchasara) [reply theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx]: @bhufmann Command handler of the open command now does the health check as well. Works well, but not sure if my reorganization is 100% correct

**Review by ebugden — CHANGES_REQUESTED** [review]
Requested changes:

- [ ] Split changes into two commits (same PR). These are distinct changes even if one is a prerequisite for the other. 
  * Commit 1: Move server health check to Open Trace command (the commit message should justify why this change was made)
  * Commit 2: Add intuitive way to open additional traces
- [ ] Nitpick: Remove the grey translucent rectangle that appears over the right half of the button when you long hover (see video below)

https://user-images.githubusercontent.com/28311615/145594722-87c804a7-7a33-472e-a243-a9b5088ce4d3.mp4

----

I can look at the code more closely after the commit is split because then it will take me less time to understand.  

I'm excited to merge this! It really is a great UX improvement. Thank you for the update @hriday-panchasara! 🌿 This new code organization makes sense to me. Thank you for the recommendation @bhufmann!

**Comment by hriday-panchasara** [comment]
> Requested changes:
> 
> * [ ]  Split changes into two commits (same PR). These are distinct changes even if one is a prerequisite for the other.
>   
>   * Commit 1: Move server health check to Open Trace command (the commit message should justify why this change was made)
>   * Commit 2: Add intuitive way to open additional traces
> * [ ]  Nitpick: Remove the grey translucent rectangle that appears over the right half of the button when you long hover (see video below)
> 
>  TraceCompassTutorialTraces.Theia-Trace.Example.Application.-.Google.Chrome.2021-12-06.09-18-40.mp4 
> I can look at the code more closely after the commit is split because then it will take me less time to understand.
> 
> I'm excited to merge this! It really is a great UX improvement. Thank you for the update @hriday-panchasara! 🌿 Also this new code organization makes sense to me. Thank you for the recommendation @bhufmann!

- Split the changes into two commits so its more readable
- I'm unable to replicate the issue with the button overlay, its appearing normal for me in light and dark themes. Might be a theia bug since we're contributing a command through its api:
Screen Shot 2021-12-10 at 10 34 05 AM
Screen Shot 2021-12-10 at 10 34 33 AM

File: theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx
Comment (ebugden) [inline theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx:88]: It's necessary to register an additional command? It's not possible to just directly use the "OpenTraceCommand"? (This is a genuine question. I don't know how commands work.)

**Review by ebugden — CHANGES_REQUESTED** [review]
Requested change:
- [ ]  Remove "fixes #193" from the "Move server health check to Open Trace command" commit. This commit is a prerequisite for fixing that issue, but it doesn't fix it. If you want to link it to the issue you could say "Required for fixing #193".

I'm being particular about commit structure and messages because it makes things much easier for future devs to understand and make changes to the code. Imagine if you're trying to understand a file in a new code base and by using "git blame" you can get clear descriptions of the purpose of each line of code from the devs that originally added them. That sounds great to me!

Apart from that, these changes look good to me!. The division of the code in each commit makes sense to me as well as the changes themselves. Thank you for taking the time to adjust the commits. Again, I'll leave it to more experienced coders to give the final approval.

**Comment by ebugden** [comment]
> * I'm unable to replicate the issue with the button overlay, its appearing normal for me in light and dark themes. Might be a theia bug since we're contributing a command through its api:
>   Screen Shot 2021-12-10 at 10 34 05 AM
>   Screen Shot 2021-12-10 at 10 34 33 AM

No problem! It's not happening for me either in the most recent version of these changes so who knows 😄

**Comment by hriday-panchasara** [comment]
> Requested change:
> 
> * [ ]   Remove "fixes #193" from the "Move server health check to Open Trace command" commit. This commit is a prerequisite for fixing that issue, but it doesn't fix it. If you want to link it to the issue you could say "Required for fixing #193".
> 
> I'm being particular about commit structure and messages because it makes things much easier for future devs to understand and make changes to the code. Imagine if you're trying to understand a file in a new code base and by using "git blame" you can get clear descriptions of the purpose of each line of code from the devs that originally added them. That sounds great to me!
> 
> Apart from that, these changes look good to me!. The division of the code in each commit makes sense to me as well as the changes themselves. Thank you for taking the time to adjust the commits. Again, I'll leave it to more experienced coders to give the final approval.

Reworded the commit message

File: theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx
Comment (hriday-panchasara) [reply theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx:88]: Since we cannot add buttons to phosphor widget headers, we're creating a toolbar command (https://theia-ide.org/docs/commands_keybindings/) to create a custom action in the widget toolbar, similar to the way we add zoom-in & zoom-out buttons to the TabBar Widget when Resources view is opened 

The OPEN_TRACE command has the label and folder icon assigned to it, and we register the command using the registerCommand function where we specify that we want it visible only for the opened traces widget. The registerCommand function also specifies callback function (OpenTraceCommand) for when its triggered.

File: theia-extensions/viewer-prototype/src/browser/trace-explorer/trace-explorer-sub-widgets/trace-explorer-placeholder-widget.tsx
Comment (bhufmann) [inline theia-extensions/viewer-prototype/src/browser/trace-explorer/trace-explorer-sub-widgets/trace-explorer-placeholder-widget.tsx:71]: I think `await` is needed here. What do you think?

File: theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx
Comment (bhufmann) [inline theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx:96]: `await` here to, right? If yes, you will need to change line 95 to add async modifier (`execute: async`)

**Review by bhufmann — CHANGES_REQUESTED** [review]

File: theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-contribution.ts
Comment (mirsky-work) [inline theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-contribution.ts:146]: Extracting that block to a separate function, called something like launchTraceServer() would improve readability - for your discretion.

**Review by mirsky-work — COMMENTED** [review]
Works well. Thanks for the contribution!

Please format the message of the commit 8fe9de8 to look like a paragraph (capital letter at the begging of a sentence, punctuations, lines capped at 80 chars, etc.).

**Comment by mirsky-work** [comment]
> * I'm unable to replicate the issue with the button overlay, its appearing normal for me in light and dark themes. Might be a theia bug since we're contributing a command through its api:

I've seen it several times in the latest patch, but can't reliably reproduce it. Anyway, it's not a big issue and shouldn't block this PR.

File: theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-contribution.ts
Comment (PatrickTasse) [inline theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-contribution.ts:60]: Looks like this method extraction belongs in the previous commit?

**Review by PatrickTasse — CHANGES_REQUESTED** [review]

**Review by ebugden — COMMENTED** [review]
Thank you for addressing the comments @hriday-panchasara! From a functional perspective I'm comfortable merging this and I'll leave it to others to give approval of the code.

**Review by ebugden — APPROVED** [review]

**Review by bhufmann — APPROVED** [review]
Looks good to me. Thanks for the contribution!

**Review by Rodrigoplp-work — APPROVED** [review]

**Review by PatrickTasse — APPROVED** [review]

**Pull request merged**
